app/models/utility/binned_normalisation_calculator.rb

```diff
@@ -0,0 +1,127 @@
+  # the bins on the child plate.
+  class BinnedNormalisationCalculator
+    include ActiveModel::Model
+    include Utility::CommonCalculations
+
+    attr_reader :config
+
+    def initialize(config)
+      @config = Utility::BinnedNormalisationConfig.new(config)
+    end
+
+    delegate :to_bigdecimal, :source_volume, :diluent_volume, :number_of_bins, :bins_template,
+             :number_decimal_places, to: :config
+
+    def normalisation_details(plate)
+      plate.wells_in_columns.each_with_object({}) do |well, details|
+        next if well.aliquots.blank?
+
+        sample_conc = to_bigdecimal(well.latest_concentration.value)
+        vol_source_reqd = compute_vol_source_reqd(sample_conc).round(number_decimal_places)
+        vol_diluent_reqd = (config.target_volume - vol_source_reqd).round(number_decimal_places)
+        amount = (vol_source_reqd * sample_conc).round(number_decimal_places)
+        dest_conc = (amount / config.target_volume).round(number_decimal_places)
+
+        details[well.location] = {
+          'vol_source_reqd' => vol_source_reqd,
+          'vol_diluent_reqd' => vol_diluent_reqd,
+          'amount_in_target' => amount,
+          'dest_conc' => dest_conc
+        }
+      end
+    end
+
+    def compute_vol_source_reqd(sample_conc)
+      vol_source_reqd = config.target_amount / sample_conc
+      vol_source_reqd = config.minimum_source_volume if vol_source_reqd < config.minimum_source_volume
+      vol_source_reqd = config.target_volume if vol_source_reqd > config.target_volume
+      vol_source_reqd
+    end
+
+    def compute_well_transfers(plate)
+      norm_details = normalisation_details(plate)
+      compute_well_transfers_hash(norm_details, plate.number_of_rows, plate.number_of_columns)
+    end
+
+    def compute_well_transfers_hash(norm_details, number_of_rows, number_of_columns)
+      conc_bins = concentration_bins(norm_details)
+      compression_reqd = compression_required?(conc_bins, number_of_rows, number_of_columns)
+      build_transfers_hash(conc_bins, number_of_rows, compression_reqd)
+    end
+
+    def compute_destination_concentrations(transfer_hash)
+      transfer_hash.values.each_with_object({}) do |dest_details, dest_hash|
+        dest_hash[dest_details['dest_locn']] = dest_details['dest_conc'].to_f
+      end
+    end
+
+    # Used by the plate presenter. Uses the concentration in the well and the plate purpose config
+    # to work out the well bin colour and number of PCR cycles.
+    def compute_presenter_bin_details(plate)
+      well_amounts = compute_well_amounts(plate)
+      compute_bin_details_by_well(well_amounts)
+    end
+
+    private
+
+    # Sorts well locations into bins based on their amounts and the configuration values.
+    def concentration_bins(norm_details)
+      conc_bins = (1..number_of_bins).each_with_object({}) { |bin_number, bins_hash| bins_hash[bin_number] = [] }
+      norm_details.each do |well_locn, details|
```

Comment:
hash = Hash.new { |h,i| h[i] = [] } IF you don't need all the keys populated regardless of content